### PR TITLE
Freemabd1/rw 15

### PR DESCRIPTION
**Pull request by freemabd1**
The pull request is for jira story RW-15. I still need to do alot to make this robust but this is the bare minimum to get connected to BigQuery.

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        // Get the results.
+        QueryResponse response = bigquery.getQueryResults(jobId);
+        QueryResult result = response.getResult();
+
+        CriteriaListResponse criteriaResponse = new CriteriaListResponse();
+        for (List<FieldValue> row : result.iterateAll()) {
+            final Criteria criteria = new Criteria();
+            criteria.setId(row.get(0).getLongValue());
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java]: I forsee this method of extracting results from queries as a bit brittle / hard to follow, since we're using indexes here instead of column names.

What may make sense: have a helper that runs the query, constructs a map of column name to index using this method:

http://googlecloudplatform.github.io/google-cloud-java/0.2.7/apidocs/com/google/cloud/bigquery/QueryResult.html#schema()

and then lets you retrieve values from rows by column name instead of index (by looking up the index.)

File: api/src/main/resources/workbench.yaml [hunk 2]
```diff
@@ -470,3 +494,50 @@ definitions:
...
+    properties:
+      id:
+        description: id of the criteria
+        type: integer
+        format: int64
+      type:
+        description: type of criteria
+        type: string
+      code:
+        description: what level of data access the user has
```
Comment (danqrodney) [inline api/src/main/resources/workbench.yaml]: This looks wrong?

File: api/src/main/resources/workbench.yaml [hunk 3]
```diff
@@ -470,3 +494,50 @@ definitions:
...
+        type: string
+      count:
+        description: est. count in the cdr
+        type: integer
+        format: int64
+      group:
+        description: specifies of child or parent
+        type: boolean
+      selectable:
+        description: specifies if use can search with
```
Comment (danqrodney) [inline api/src/main/resources/workbench.yaml]: Which criteria would we not be able to search on?

File: api/src/main/resources/workbench.yaml [hunk 4]
```diff
@@ -470,3 +494,50 @@ definitions:
...
+        type: string
+      name:
+        description: description of criteria
+        type: string
+      count:
+        description: est. count in the cdr
+        type: integer
+        format: int64
+      group:
+        description: specifies of child or parent
```
Comment (danqrodney) [inline api/src/main/resources/workbench.yaml]: "child" or "parent"? 

Is the idea that nodes which are "parent" are not selectable, and are just containers for children?

File: api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                        "SELECT id, " +
+                                "type, " +
+                                "code, " +
+                                "name, " +
+                                "est_count, " +
+                                "is_group, " +
+                                "is_selectable, " +
+                                "domain_id " +
+                                "FROM `pmi-drc-api-test.synpuf.icd9_criteria` " +
+                                "where parent_id = 0 " +
```
[the remark this answers is not in the document]
Comment (freemabd1) [reply api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java]: Right.. that will definitely change.

File: api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                                "order by id asc;")
+                        // Use standard SQL syntax for queries.
+                        // See: https://cloud.google.com/bigquery/sql-reference/
+                        .setUseLegacySql(false)
+                        .build();
+
+        // Create a job ID so that we can safely retry.
```
[the remark this answers is not in the document]
Comment (freemabd1) [reply api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java]: Agreed.. i think the link about also uses query()

File: api/src/main/resources/workbench.yaml (hunk 2 above)
Comment (freemabd1) [reply api/src/main/resources/workbench.yaml]: oops

File: api/src/main/resources/workbench.yaml (hunk 4 above)
Comment (freemabd1) [reply api/src/main/resources/workbench.yaml]: some are not selectable and some are depending on the specific criteria tree, so we use a combination of isGroup and IsSelectable to determine what is a parent or leaf and is selectable

Comment (danqrodney) [reply api/src/main/resources/workbench.yaml]: So "group" = "true" means there are children, and "selectable" = "true" means you can query on it?

Comment (freemabd1) [reply api/src/main/resources/workbench.yaml]: yes

File: api/src/main/resources/workbench.yaml (hunk 3 above)
Comment (freemabd1) [reply api/src/main/resources/workbench.yaml]: see above

File: api/build.gradle [hunk 7]
```diff
@@ -120,6 +120,7 @@ dependencies {
   compile 'mysql:mysql-connector-java:+'
   compile 'com.google.cloud.sql:mysql-socket-factory:+'
   compile 'com.google.appengine:appengine:+'
+  compile 'com.google.appengine:appengine-api-1.0-sdk:+'
```
[the remark this answers is not in the document]
Comment (freemabd1) [reply api/build.gradle]: So i get this error without that dependency:  java.lang.RuntimeException: Google App Engine runtime detected (the environment variable "com.google.appengine.runtime.version" is set), but unable to resolve appengine-sdk classes. For more details see https://github.com/GoogleCloudPlatform/google-cloud-java/blob/master/APPENGINE.md

[the remark this answers is not in the document]
Comment (danqrodney) [reply api/build.gradle]: Interesting. Thanks for the info!

File: api/libproject/devstart.rb [hunk 8]
```diff
@@ -58,6 +59,19 @@ def dev_up(args)
   common.run_inline %W{docker-compose up -d db}
   common.status "Running database migrations..."
   common.run_inline %W{docker-compose run db-migration}
+  common.status "Creating service account for API..."
+  iam_account = "bigqueryservice@all-of-us-workbench-test.iam.gserviceaccount.com"
```
[the remark this answers is not in the document]
Comment (freemabd1) [reply api/libproject/devstart.rb]: we can certainly do that. I don't have access to create a service account under @appspot.gserviceaccount.com. I can currently only create under all-of-us-workbench-test.iam.gserviceaccount.com and @all-of-us-ehr-dev.iam.gserviceaccount.com

[the remark this answers is not in the document]
Comment (danqrodney) [reply api/libproject/devstart.rb]: You don't need to create a service account, it already exists. (It's the SA that our deployed AppEngine app in test uses.)

You will need a key for it, but you are an Owner on all-of-us-workbench-test, so you should have permissions to do that in cloud console or on the command line (like this).

File: api/libproject/devstart.rb [hunk 9]
```diff
@@ -58,6 +59,19 @@ def dev_up(args)
...
+  common.run_inline %W{
+    gcloud iam service-accounts keys create sa-key.json --iam-account #{iam_account}
```
[the remark this answers is not in the document]
Comment (freemabd1) [reply api/libproject/devstart.rb]: I guess i'm still a little confused about this and the above comments. Maybe we can discuss over a call?

[the remark this answers is not in the document]
Comment (danqrodney) [reply api/libproject/devstart.rb]: Definitely. Let's chat on Monday.

File: api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+
+@RestController
+public class CohortBuilderController implements CohortBuilderApiDelegate {
+
+    private static final Logger log = Logger.getLogger(CohortBuilderController.class.getName());
+
+    public static final String CRTIERIA_QUERY = "SELECT id, type, code, name, est_count, is_group, is_selectable, domain_id\n"
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java]: There are some long lines in here. We should wrap them.

Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java]: CRTIERIA -> CRITERIA

File: api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+            if (response.getExecutionErrors().size() != 0) {
+                firstError = response.getExecutionErrors().get(0).getMessage();
+            }
+            throw new RuntimeException(firstError);
+        }
+
+        return response.getResult();
+    }
+
+    protected Map<String, Integer> getResultMapper(QueryResult result) {
```
Comment (danqrodney) [inline api/src/main/java/org/pmiops/workbench/api/CohortBuilderController.java]: Eventually we'll probably want to move this and some of the code above into some BigQuery utils class. (Fine to leave it here for now.)
